control-plane/agents/src/bin/core/volume/snapshot_operations.rs

```diff
@@ -155,9 +175,52 @@ impl ResourceLifecycleWithLifetime for OperationGuardArc<VolumeSnapshot> {
         registry: &Registry,
         _request: &Self::Destroy<'_>,
     ) -> Result<(), SvcError> {
-        self.start_destroy_by(registry, &()).await?;
-        let result = Ok(());
-        self.complete_destroy(result, registry).await
+        // Get the volume snapshot persisted info.
+        let volume_snapshot = self.lock().clone();
+
+        // Create a map to track the transactions upon deletion.
+        let mut updated_transactions: HashMap<SnapshotTxId, Vec<ReplicaSnapshot>> = HashMap::new();
+        let mut error_occurred = false;
+
+        // Iterate over current transactions and destroy each replica snapshot in a transaction,
+        // store the transactions and replica snapshots whose deletion failed.
+        for (txn_id, snap_list) in volume_snapshot.metadata().transactions() {
+            for snap in snap_list {
+                let mut cloned_snap = snap.clone();
+                if let Err(err) = ReplicaSnapshotDestroy::new(cloned_snap.clone())
+                    .execute_destroy(registry)
+                    .await
+                {
+                    if !matches!(err, SvcError::NotFound { .. }) {
+                        // If it was an error mark the spec as deleting and push it in a map.
+                        cloned_snap.set_status_deleting();
+                        updated_transactions
+                            .entry(txn_id.clone())
+                            .or_insert_with(Vec::new)
+                            .push(cloned_snap);
+                        error_occurred = true;
+                    }
+                }
+            }
+        }
+
+        if error_occurred {
+            // If errors were encountered then update the persisted spec with updated transactions.
+            let spec_clone = self
+                .start_update(
+                    registry,
+                    &volume_snapshot,
+                    VolumeSnapshotOperation::UpdateTransactions(updated_transactions),
+                )
+                .await?;
+            self.complete_update(registry, Ok(()), spec_clone).await?;
+        } else {
+            // If there we no errors remove the persisted spec and complete destroy.
+            self.start_destroy(registry).await?;
+            self.complete_destroy(Ok(()), registry).await?;
```

Comment:
hmm, wondering I think should always start_destroy otherwise we'll keep saying the snapshot is created which doesn't seem right?

Reply:
Incorporated this as a part of the Operation.